<a href="https://colab.research.google.com/github/sovikapoudel/Data-Visualiation-Spring-2022/blob/main/Quiz17.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Grab the iris dataset. Answer the following questions:**

In [4]:
import pandas as pd

iris_df = pd.read_csv('https://raw.githubusercontent.com/nurfnick/Data_Viz/main/iris.csv')
iris_df

SepalLength  SepalWidth  PedalLength  PedalWidth           Class
0            5.1         3.5          1.4         0.2     Iris-setosa
1            4.9         3.0          1.4         0.2     Iris-setosa
2            4.7         3.2          1.3         0.2     Iris-setosa
3            4.6         3.1          1.5         0.2     Iris-setosa
4            5.0         3.6          1.4         0.2     Iris-setosa
..           ...         ...          ...         ...             ...
145          6.7         3.0          5.2         2.3  Iris-virginica
146          6.3         2.5          5.0         1.9  Iris-virginica
147          6.5         3.0          5.2         2.0  Iris-virginica
148          6.2         3.4          5.4         2.3  Iris-virginica
149          5.9         3.0          5.1         1.8  Iris-virginica

[150 rows x 5 columns]


##2 Does the direction of the shift remain the same if you groupby Class?
##3 Gather the mean, median, count and standard deviation of all columns when grouped by Class.

In [26]:
import statistics



In [27]:
from numpy.ma.core import mean
mean(iris_df.SepalWidth)

3.0540000000000003

In [28]:
new_df= iris_df.SepalWidth.astype('int') 
new_df

0      3
1      3
2      3
3      3
4      3
      ..
145    3
146    2
147    3
148    3
149    3
Name: SepalWidth, Length: 150, dtype: int64

In [29]:
mean(new_df)

2.6466666666666665

Here, you can see the int function gave the whole number and remove the decimal and I think it is not best practice when it comes to sensitive data because it will just chop off the decimal value without rounding them. For example, 3.9 would be 3

In [30]:
rnew_df=round(iris_df.SepalWidth).astype('int')
rnew_df

0      4
1      3
2      3
3      3
4      4
      ..
145    3
146    2
147    3
148    3
149    3
Name: SepalWidth, Length: 150, dtype: int64

In [31]:
mean(rnew_df)

3.033333333333333

I think this would be the best option as it rouded up and gave integer value.

##1. Does converting SepalLength to integer increase or decrease the mean?

It does impact the mean value as you can see on my output, when I just take the average of float value provided on the data set, I got 3.0540000000000003 but when I change the type to integer then I got 2.6466666666666665 which seems significantly low than the float mean, this is beacuse the digit were not rounded up when changing into integer. However, when I change thhe data type after rounding them up I got the value of 3.033333333333333 which is close to the first value.

####So I understanding is, if we do not round up, then changing it into int type would decrease the mean as it ignores the decimal value.

####And If we round up and find the mean then it depends on the decimal value whether it is less than .5 or greater. If the majority of numbers are greater than .5 decimal then it would increase the mean and if less it would decrease.

###2. Does the direction of the shift remain the same if you groupby Class?

Let's first apply shift function


In [45]:
df_test=iris_df.shift()
df_test


SepalLength  SepalWidth  PedalLength  PedalWidth           Class
0            NaN         NaN          NaN         NaN             NaN
1            5.1         3.5          1.4         0.2     Iris-setosa
2            4.9         3.0          1.4         0.2     Iris-setosa
3            4.7         3.2          1.3         0.2     Iris-setosa
4            4.6         3.1          1.5         0.2     Iris-setosa
..           ...         ...          ...         ...             ...
145          6.7         3.3          5.7         2.5  Iris-virginica
146          6.7         3.0          5.2         2.3  Iris-virginica
147          6.3         2.5          5.0         1.9  Iris-virginica
148          6.5         3.0          5.2         2.0  Iris-virginica
149          6.2         3.4          5.4         2.3  Iris-virginica

[150 rows x 5 columns]

Now, applying shift after grouping them by class

In [44]:
col= iris_df.groupby('Class')
col.shift()

SepalLength  SepalWidth  PedalLength  PedalWidth
0            NaN         NaN          NaN         NaN
1            5.1         3.5          1.4         0.2
2            4.9         3.0          1.4         0.2
3            4.7         3.2          1.3         0.2
4            4.6         3.1          1.5         0.2
..           ...         ...          ...         ...
145          6.7         3.3          5.7         2.5
146          6.7         3.0          5.2         2.3
147          6.3         2.5          5.0         1.9
148          6.5         3.0          5.2         2.0
149          6.2         3.4          5.4         2.3

[150 rows x 4 columns]

I might not be doing what question have asked for but The direction of the shift does not remains same if groupby class in terms of column number. Once it

###3 Gather the mean, median, count and standard deviation of all columns when grouped by Class.

In [23]:
iris_df.groupby('Class').SepalWidth.agg(['mean','median','count', 'std'])

mean  median  count       std
Class                                          
Iris-setosa      3.418     3.4     50  0.381024
Iris-versicolor  2.770     2.8     50  0.313798
Iris-virginica   2.974     3.0     50  0.322497